// astro/dist/core/build/plugins/plugin-middleware.js
import { MIDDLEWARE_PATH_SEGMENT_NAME } from "../../constants.js";
import { addRollupInput } from "../add-rollup-input.js";
const MIDDLEWARE_MODULE_ID = "@astro-middleware";
function vitePluginMiddleware(opts, _internals) {
  return {
    name: "@astro/plugin-middleware",
    options(options) {
      if (opts.settings.config.experimental.middleware) {
        return addRollupInput(options, [MIDDLEWARE_MODULE_ID]);
      }
    },
    async resolveId(id) {
      if (id === MIDDLEWARE_MODULE_ID && opts.settings.config.experimental.middleware) {
        const middlewareId = await this.resolve(
          `${opts.settings.config.srcDir.pathname}/${MIDDLEWARE_PATH_SEGMENT_NAME}`
        );
        if (middlewareId) {
          return middlewareId.id;
        }
      }
    }
  };
}
function pluginMiddleware(opts, internals) {
  return {
    build: "ssr",
    hooks: {
      "build:before": () => {
        return {
          vitePlugin: vitePluginMiddleware(opts, internals)
        };
      }
    }
  };
}
export {
  MIDDLEWARE_MODULE_ID,
  pluginMiddleware,
  vitePluginMiddleware
};
